src/UserSpaceInstrumentation/AllocateInTracee.cpp

```diff
@@ -0,0 +1,194 @@
+  auto result_memory_region = GetFirstExecutableMemoryRegion(pid, &address_start, &address_end);
+  if (result_memory_region.has_error()) {
+    return ErrorMessage(absl::StrFormat("Failed to find executable memory region: \"%s\"",
+                                        result_memory_region.error().message()));
+  }
+
+  // Backup first 8 bytes.
+  const uint64_t backup_module_code = ptrace(PTRACE_PEEKDATA, pid, address_start, NULL);
+  if (errno) {
+    return ErrorMessage(absl::StrFormat("Failed to PTRACE_PEEKDATA with errno %d: \"%s\"", errno,
+                                        SafeStrerror(errno)));
+  }
+
+  // Write `syscall` into memory. Machine code is `0x0f05`.
+  WriteBytesIntoTraceesMemory(pid, address_start, std::vector<uint8_t>{0x0f, 0x05});
+
+  // Move instruction pointer to the `syscall` and fill registers with parameters.
+  RegisterState registers_for_syscall = original_registers;
+  registers_for_syscall.GetGeneralPurposeRegisters()->x86_64.rip = address_start;
+  registers_for_syscall.GetGeneralPurposeRegisters()->x86_64.rax = syscall;
+  // Register list for arguments can be found e.g. in the glibc wrapper:
+  // https://github.com/bminor/glibc/blob/master/sysdeps/unix/sysv/linux/x86_64/syscall.S#L30
+  registers_for_syscall.GetGeneralPurposeRegisters()->x86_64.rdi = arg_0;
+  registers_for_syscall.GetGeneralPurposeRegisters()->x86_64.rsi = arg_1;
+  registers_for_syscall.GetGeneralPurposeRegisters()->x86_64.rdx = arg_2;
+  registers_for_syscall.GetGeneralPurposeRegisters()->x86_64.r10 = arg_3;
+  registers_for_syscall.GetGeneralPurposeRegisters()->x86_64.r8 = arg_4;
+  registers_for_syscall.GetGeneralPurposeRegisters()->x86_64.r9 = arg_5;
+  auto result_restore_registers = registers_for_syscall.RestoreRegisters();
+  if (result_restore_registers.has_error()) {
+    return ErrorMessage(absl::StrFormat("Failed to set registers with syscall parameters: \"%s\"",
+                                        result_restore_registers.error().message()));
+  }
+
+  // Single step to execute the syscall.
+  if (ptrace(PTRACE_SINGLESTEP, pid, 0, 0) != 0) {
+    return ErrorMessage("Failed to execute syscall with PTRACE_SINGLESTEP.");
+  }
+  int status = 0;
+  pid_t waited = waitpid(pid, &status, 0);
+  if (waited != pid || !WIFSTOPPED(status) || WSTOPSIG(status) != SIGTRAP) {
+    return ErrorMessage("Failed to wait for PTRACE_SINGLESTEP to execute.");
+  }
+
+  // Return value of syscalls is in rax.
+  RegisterState return_value;
+  auto result_backup_return_value = return_value.BackupRegisters(pid);
+  if (result_backup_return_value.has_error()) {
+    return ErrorMessage(absl::StrFormat("Failed to get registers with mmap result: \"%s\"",
+                                        result_backup_return_value.error().message()));
+  }
+  const uint64_t result = return_value.GetGeneralPurposeRegisters()->x86_64.rax;
+  // Syscalls return -4095, ..., -1 on failure.
+  const int64_t result_as_int = reinterpret_cast<uint64_t>(result);
```

Comment:
I think we somewhat decided to prefer `absl::bit_cast` over `reinterpret_cast`, it would apply to this case.

Reply:
done